pyutilib/misc/timing.py

```diff
@@ -160,3 +160,305 @@ def start(self):
+            ------------------------------------------------------------------------
+            aa                 6.60e-02             50       1.32e-03           57.3%
+            ab                 4.84e-02             10       4.84e-03           42.0%
+            other              8.00e-04            N/A            N/A            0.7%
+        b                  1.39e-02             10       1.39e-03           10.8%
+        other              3.49e-04            N/A            N/A            0.3%
+    >>>
+    >>>
+    >>> print('a total time: ', timer.get_total_time('all.a'))
+    a total time:  0.11518359184265137
+    >>> 
+    >>> print('ab num calls: ', timer.get_num_calls('all.a.ab'))
+    ab num calls:  10
+    >>> 
+    >>> print('aa % time: ', timer.get_relative_percent_time('all.a.aa'))
+    aa % time:  57.28656738043737
+    >>>
+    >>> print('aa % of total time: ', timer.get_total_percent_time('all.a.aa'))
+    aa % of total time:  50.96888018003749
+
+    Internal Workings
+    -----------------
+    The HierarchicalTimer use a stack to track which timers are active
+    at any point in time. Additionally, each timer has a dictionary of
+    timers for its children timers. Consider
+
+    >>> timer = HierarchicalTimer()
+    >>> timer.start_increment('all')
+    >>> timer.start_increment('a')
+    >>> timer.start_increment('aa')
+
+    After the above code is run, self.stack will be ['all', 'a', 'aa']
+    and self.timers will have one key, 'all' and one value which will
+    be a _HierarchicalHelper. The _HierarchicalHelper has its own timers dictionary: 
+    
+    {'a': _HierarchicalHelper}
+
+    and so on. This way, we can easily access any timer with something
+    that looks like the stack. The logic is recursive (although the
+    code is not).
+
+    """
+    def __init__(self):
+        self.stack = list()
+        self.timers = dict()
+
+    def _get_timer(self, identifier, should_exist=False):
+        """
+
+        This method gets the timer associated with the current state
+        of self.stack and the specified identifier.
+
+        Parameters
+        ----------
+        identifier: str
+            The name of the timer
+        should_exist: bool
+            The should_exist is True, and the timer does not already
+            exist, an error will be raised. If should_exist is False, and
+            the timer does not already exist, a new timer will be made.
+        
+        Returns
+        -------
+        timer: _HierarchicalHelper
+
+        """
+        parent = self._get_timer_from_stack(self.stack)
+        if identifier in parent.timers:
+            return parent.timers[identifier]
+        else:
+            if should_exist:
+                raise RuntimeError('Could not find timer {0}'.format('.'.join(self.stack + [identifier])))
+            parent.timers[identifier] = _HierarchicalHelper()
+            return parent.timers[identifier]
+
+    def start_increment(self, identifier):
```

Comment:
The TicTocTimer uses `start()` and `stop()`.  Can we make the two APIs consistent?